plugin/manager.go

```diff
@@ -20,82 +22,75 @@ func init() {
 
 // Plugin is a generic struct for storing info about a plugin.
 type Plugin struct {
-	Path          string
-	Client        *goplugin.Client
-	Reacter       Reacter
-	Subscriptions []Subscription
+	Path    string
+	Reacter Reacter
 }
 
 // Manager handles lifecycle of plugin management.
 type Manager struct {
-	Plugins []*Plugin
-	Log     *zap.Logger
+	Reacters []*Plugin
+	Log      *zap.Logger
 }
 
 // NewManager creates new Manager.
-func NewManager(paths []string, log *zap.Logger) *Manager {
-	plugins := []*Plugin{}
+func NewManager(paths []string, log *zap.Logger) (*Manager, error) {
+	reacters := []*Plugin{}
 	logger := Hclog2ZapLogger{Zap: log}
+
 	for _, path := range paths {
 		client := goplugin.NewClient(&goplugin.ClientConfig{
-			HandshakeConfig: handshakeConfig,
+			HandshakeConfig: shared.Handshake,
 			Plugins:         pluginMap,
 			Cmd:             exec.Command(path),
 			Logger:          logger.Named("PluginManager"),
+			Managed:         true,
 		})
 
-		plugins = append(plugins, &Plugin{
-			Client: client,
-			Path:   path,
-		})
-	}
-
-	return &Manager{
-		Plugins: plugins,
-		Log:     log,
-	}
-}
-
-// Connect connects to plugins.
-func (m *Manager) Connect() error {
-	for _, plugin := range m.Plugins {
-		rpcClient, err := plugin.Client.Client()
+		rpcClient, err := client.Client()
 		if err != nil {
-			return err
+			return nil, err
 		}
 
 		// Request the plugin
-		raw, err := rpcClient.Dispense("subscriber")
+		raw, err := rpcClient.Dispense("reacter")
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		plugin.Reacter = raw.(*Subscriber)
-		plugin.Subscriptions = plugin.Reacter.Subscriptions()
+		switch raw.(type) {
+		case *ReacterClient:
+			reacters = append(reacters, &Plugin{
+				Path:    path,
+				Reacter: raw.(*ReacterClient),
+			})
+		default:
+			panic("plugin unrecognized")
+		}
 	}
 
-	return nil
+	return &Manager{
+		Reacters: reacters,
+		Log:      log,
+	}, nil
 }
 
 // Kill disconnects plugins and kill subprocesses.
 func (m *Manager) Kill() {
-	for _, plugin := range m.Plugins {
-		plugin.Client.Kill()
-	}
+	goplugin.CleanupClients()
 }
 
 // React call all plugins' React method. It returns when the first error is returned by a plugin.
 func (m *Manager) React(event *event.Event) error {
-	for _, plugin := range m.Plugins {
-		for _, subscription := range plugin.Subscriptions {
-			if subscription.EventType == event.EventType {
+	for _, plugin := range m.Reacters {
+		for _, sub := range plugin.Reacter.Subscriptions() {
+			if sub.EventType == event.EventType {
 				err := plugin.Reacter.React(*event)
 				if err != nil {
 					m.Log.Debug("Plugin returned error.",
 						zap.String("plugin", plugin.Path),
 						zap.Error(err),
-						zap.String("subscriptionType", string(subscription.Type)))
-					if subscription.Type == Sync {
+						zap.String("subscriptionType", string(sub.Type)))
+					if sub.Type == subscription.TypeSync {
```

Comment:
I'm not a fan of using `Type` in all of these places...it's quite confusing given how close this is to the keyword itself. Is it possible to define `type Type string` in `subscription` package to be `SubscriptionType` to help differentiate, or something slightly shorter that that?

Reply:
The type is a string because it's used in config API as a Type field in subscription so it has to be human readable. 

I will revert this change to the previous stage. So there will be separate consts for that.